services/rpcfilters/latest_block_changed_event.go

```diff
@@ -64,21 +103,18 @@ func (e *latestBlockChangedEvent) numberOfSubscriptions() int {
 }
 
 func (e *latestBlockChangedEvent) processLatestBlock(latestBlock blockInfo) {
-	e.latestBlockMu.Lock()
-	defer e.latestBlockMu.Unlock()
-
-	// if we received the same or an older block than we already have, ignore it.
-	if latestBlock.Number().Cmp(e.previousLatestBlock.Number()) <= 0 {
+	// if we received the hash we already received before, don't add it
+	if e.reportedBlocks.Has(latestBlock.Hash) {
```

Comment:
Don't we have a possible race here? In case `Has()` returns false but it's added before `Add()`? So better one method `AddOnce(common.Hash) bool)`.

Reply:
Bad formulated, yes. It's about the locking of individual `ringArray` methods, but when you test with `Has()`, get a false, and then use `Add()` the hash to add already could be add by another goroutine having a reference to the `ringArray`.

May be that this situation doesn't happen in your use case right now but then the local `ringArray` doesn't need a mutex.

Reply:
ah, you are right, I potentially need to lock for the whole check/add cycle.

Reply:
@themue fixed! thanks for pointing this out!